docs/proposals/emptydir-pv.md

```diff
@@ -0,0 +1,151 @@
+```
+
+### Proposed flow
+
+At a high level:
+* Kubelet watches volumes (new)
+* Populator observes pod attached to emptydir PV and adds volume spec to dsw (unchanged)
+* Reconciler notices new volume spec in dsw not in asw, invokes new plugin (unchanged)
+* New plugin attaches, mounts (new)
+* Populator observes pod delete (unchanged)
+* Populator deletes pod from dsw volume map, keeps volume (new, today it deletes volume from map when no pods are using it)
+* Reconciler notices volume in asw with no matching pod in dsw, invokes new plugin (unchanged)
+* New plugin unmounts (new)
+* ...
+* Kubelet notices pv deletion (new)
+* Populator deletes volume from map (new)
+* Reconciler notices volume in asw with no matching volume in dsw, invokes new plugin (unchanged)
+* New plugin detaches (new)
+
+### New storage plugin
+
+We will write an attachable volume plugin that wraps emptydir. The directory is created and deleted in the Attach/Detach stage, BUT these operations are carried out through the same functions that implement emptydir Mount/Unmount. Attach must be idempotent, after creating the `/var/lib/kubelet/volumes/pv-uuid` directory, we record the volume as attached through the node status. Subsequent calls to Attach will no-op. Detach will only get called when the volume is removed from the desired state of the world, which by our design won't happen till the PV is deleted. The permissions on this directory will be the same as emptydir (i.e 777).
+
+This storage plugin will respond "yes" when asked if it owns a volume spec that contains a PV + `sticky-host` annotation, and "no" in every other case. (TODO: do we need to bind mount the directory for every pod?).
+
+## Example
+
+```yaml
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  name: test
+  annotations:
+    # Set by scheduler or admin
+    volume.alpha.kubernetes.io/sticky-host: ""
+spec:
+  capacity:
+    storage: 10Mi
+  accessModes:
+    - ReadWriteMany
+  emptyDir: {}
+---
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: test
+  annotations:
+    # Mentioned for the sake of clarity
+    volume.alpha.kubernetes.io/storage-class: free
+spec:
+  accessModes: [ "ReadWriteMany" ]
+  resources:
+    requests:
+      storage: 10Mi
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  generateName: test-
+spec:
+  terminationGracePeriodSeconds: 0
+  containers:
+  - name: busybox
+    image: busybox
+    command:
+    - sh
+    - -c
+    - touch /data/`hostname`; while true; do sleep 3600; done
+    volumeMounts:
+    - name: data
+      mountPath: /data
+  volumes:
+  - name: data
+    persistentVolumeClaim:
+      claimName: test
+```
+
+Creating this will give you a pod pinned to the node the scheduler initally picks, with a directory mounted at /data backed by `/var/lib/kubelet/volumes/pv-uuid`. Creating the same thing without the PV section will invoke a dynamic storage provisioner to create an identical PV.
```

Comment:
In the PV spec it's the persistentVolumeSource, that now says "emtpyDir" (https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L304). Currently emptyDir is not one of the allowed sources (https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L240), I'm proposing changing this. 

Without the pv spec it's the `volume.alpha.kubernetes.io/storage-class: free` that invokes the "free" dyanmic provisioner to create the pv with `emptydir: {}`.